azure-iot-device/azure/iot/device/common/transport/pipeline_stages_base.py

```diff
@@ -611,3 +613,62 @@ def on_connected(self):
     def on_disconnected(self):
         self.connected = False
         PipelineStage.on_disconnected(self)
+
+
+class CoordinateRequestAndResponse(PipelineStage):
+    """
+    Pipeline stage which is responsible for coordinating IotRequestAndResponse operations.  For each
+    IotRequestAndResponse operation, this stage passes down a SendIotRequest operation and waits for
+    an IotResponseEvent event.  All other events are passed down unmodified.
+    """
+
+    def __init__(self):
+        super(CoordinateRequestAndResponse, self).__init__()
+        self.pending_responses = {}
+
+    def _run_op(self, op):
+        if isinstance(op, pipeline_ops_base.IotRequestAndResponse):
+            # Convert IotRequestAndResponse operation into a SendIotRequest operation
+            # and send it down.  A lower level will convert the SendIotRequest into an
+            # actual transport operation.  The IotRequestAndResponse operation will be
+            # completed when the corresponding IotResponse event is received in this stage.
+
+            request_id = str(uuid.uuid4())
+
+            def on_send_request_done(send_request_op):
+                if send_request_op.error:
+                    op.error = send_request_op.error
+                    del (self.pending_responses[request_id])
+                else:
+                    # request sent.  Nothing to do except wait for the response
+                    pass
+
+            new_op = pipeline_ops_base.SendIotRequest(
+                verb=op.verb,
+                resource=op.resource,
+                request_body=op.request_body,
+                request_id=request_id,
+                callback=on_send_request_done,
+            )
+            self.continue_op(new_op)
+
+        else:
+            self.continue_op(op)
+
+    def _handle_pipeline_event(self, event):
+        if isinstance(event, pipeline_events_base.IotResponseEvent):
+            # match IotResponseEvent events to the saved dictionary of IotRequestAndResponse
+            # operations which have not received responses yet.  If the operation is found,
+            # complete it.
+
+            if event.request_id in self.pending_responses:
+                op = self.pending_responses[event.request_id]
+                del (self.pending_responses[event.request_id])
+                op.status_code = event.status_code
+                op.response_body = event.response_body
+            else:
+                # OPEN ISSUE: what to do if we don't have a matching $rid?  Drop it, error, i
```

Comment:
>OPEN ISSUE [](start = 18, length = 10)

is this response important ? like will anyone do anything with this response ? In that case it may be passed up.
This must be coming to an rid that was sent in the past. 
I would say that just log an error and log out the response , in a way drop it.  
Error may be a bit overboard.